File: html/src/app/piece.ts
import { BLANK, BLACK, WHITE, COLUMNS, ROWS, SQUARE_W, SQUARE_H } from './constants';

export class Piece {
  position = 0;
  color = BLANK;
  frame = BLANK;

  constructor() {
    this.setPosition(0);
    this.setColor(BLANK);
    this.setFrame(BLANK);
  }

  setPosition(newPosition: number): boolean {
    if (newPosition < COLUMNS * ROWS) {
      this.position = newPosition;
      return true;
    }
    return false;
  }

  getPosition(): number {
    return this.position;
  }

  setColor(newColor: number): void {
    if (newColor === BLACK) {
      this.color = BLACK;
    } else if (newColor === WHITE) {
      this.color = WHITE;
    } else {
      this.color = BLANK;
    }
  }

  getColor(): number {
    return this.color;
  }

  setFrame(newFrame: number): void {
    if (newFrame === BLACK) {
      this.frame = BLACK;
    } else if (newFrame === WHITE) {
      this.frame = WHITE;
    } else {
      this.frame = BLANK;
    }
  }

  getFrame(): number {
    return this.frame;
  }

  getPixel(): { x: number; y: number } {
    const x = this.position % COLUMNS;
    const y = Math.floor(this.position / ROWS);
    return { x: x * SQUARE_W + 1, y: y * SQUARE_H + 1 };
  }
}

File: piece.cpp
/***************************************************************************
                          othellocanvasview.cpp  -  description
                             -------------------
    begin                : Sat May 3 2003
    copyright            : (C) 2003 by Ilmari Heikkinen
    email                : irheikki@cc.hut.fi
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "piece.h"

#include <qcanvas.h>
#include <qstring.h>
#include <qimage.h>

Piece::Piece( QCanvasPixmapArray *parray, QCanvas *canvas ) : QCanvasSprite( parray, canvas ), color(BLANK)
{
  setZ( 1 );
  setAnimated( false );     // disallow animation
}

Piece::~Piece()
{}

void Piece::flip()
{
  if (color == BLACK) {
      setFrame( WHITE );   
  } else {
      setFrame( BLACK );
  }
  color = (color == BLACK) ? WHITE : BLACK;
}

// returns true if successful (new position within limits), otherwise false
bool Piece::setPosition(unsigned int new_position)
{
  if (new_position < (COLUMNS*ROWS)) {
      position = new_position;
      // move piece on board, x * square_width, y * square_height
      double x = (position % COLUMNS);
      double y = (position / ROWS);

      move(x*SQUARE_W+1, y*SQUARE_H+1);
      setZ( position );
      return true;
  }
  return false;
}

int Piece::rtti() const {
  return PIECE_RTTI;
}

unsigned int Piece::getPosition()
{
  return position;
}

void Piece::setColor( unsigned int new_color )
{
  if (new_color == BLACK) {
    color = BLACK;
  } else if (new_color == WHITE) {
    color = WHITE;
  } else {
    color = BLANK;
  }
}

unsigned int Piece::getColor()
{
  return color;
}
